File: 1013-fibonacci-number/fibonacci-number.java
class Solution {
    public int fib(int n) {
        Map<Integer, Integer> map = new HashMap<>();
        return dp(n, map);
    }

    public int dp(int n, Map<Integer, Integer> map) {
        if (n <= 0) return 0;
        if (n == 1) return 1;
        if (map.containsKey(n)) return map.get(n);
        
        int v1 = dp(n-1, map); int v2 = dp(n-2, map);

        map.put(n, v1 + v2);
        return v1+v2;
    }
}

File: 1013-fibonacci-number/fibonacci-number.swift
class Solution {
    func fib(_ n: Int) -> Int {
        if n == 0 { return 0 }
        if n == 1 { return 1 }
        var arr: [Int] = Array(repeating: 0, count: n+1)
        arr[0] = 0
        arr[1] = 1

        for i in 2...n {
            arr[i] = arr[i-1] + arr[i-2]
        }
        return arr[n]
    }
}